contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,158 @@
+
+## Non-Goals
+
+* Define how CSI volume driver should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume drivers deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource will be automatically be installed on Kubernetes clusters.
+
+Upon deployment a driver must create a new custom resource object.
+
+## Design Details
+
+### `CSIDriver` Object
+
+#### Proposed API
+
+```go
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+//
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+    metav1.TypeMeta   `json:",inline"`
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
+
+    // Specification describing the CSI volume driver and any custom
+    // configuration for it.
+	Spec CSIDriverSpec `json:"spec"`
+
+	// Status of the CSI volume driver.
+	Status CSIDriverStatus `json:"status,omitempty"`
+}
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of CSIDrivers
+	Items []CSIDriver `json:"items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // Driver indicates the name of the CSI driver that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume driver requires an attach operation (because it
+    // implements the CSI ControllerPublishVolume() method), and that Kubernetes
+    // should call attach and wait for any attach operation to complete before
+    // proceeding to mounting.
+    // If value is not specified, default is true -- meaning attach will be
+    // called.
+    // +optional
+    AttachRequired *bool `json:"attachRequired"`
+    
+    // Indicates this CSI volume driver requires additional pod information
+    // (like podName, podUID, etc.) during mount operations.
+    // If this is set to true, Kubelet will pass pod information as
+    // VolumeAttributes in the CSI NodePublishVolume() calls.
+    // If value is not specified, default is false -- meaning pod information
+    // will not be passed on mount. 
+    // +optional
+    PodInfoRequiredOnMount *bool `json:"podInfoRequiredOnMount"`
+}
+
+// CSIDriverStatus is the status of a CSIDriver.
+type CSIDriverStatus struct {
+    // Indicates the volume driver has been successfully deployed on the cluster
+    // and is ready to use.
+	Ready bool `json:"ready"`
```

Comment:
It's still unclear to me what "Ready" means.  "the volume driver has been successfully deployed" - so all containers (DaemonSet with node components and Deployment with controller components) are running? What one of them fails, who sets `Ready=false`? What if I add a new node, shall something set it to `Ready=false`, wait until the DaemonSet runs a new pod with node components there and set it to `Ready=true`?

Reply:
Will remove